### PR TITLE
hw30: - done;

### PR DESCRIPTION
**Pull request by jah701**

### CODE REVIEW

File: src/main/java/internet/shop/dao/jdbc/ProductDaoJdbcImpl.java [hunk 1]
```diff
@@ -0,0 +1,106 @@
...
+            throw new RuntimeException("Can't create product", e);
+        }
+    }
+
+    @Override
+    public Optional<Product> get(Long id) {
+        Product product = null;
+        try (Connection connection = ConnectionUtil.getConnection()) {
+            PreparedStatement statement = connection.prepareStatement(
+                    "SELECT * FROM products WHERE deleted = 0 AND product_id = ?");
```
Comment (AnaLanda) [inline src/main/java/internet/shop/dao/jdbc/ProductDaoJdbcImpl.java]: Not sure about this, but maybe we should have semicolons at the end of the SQL query, like this: `...product_id=?;");` ?

File: src/main/java/internet/shop/dao/jdbc/ProductDaoJdbcImpl.java [hunk 2]
```diff
@@ -0,0 +1,106 @@
...
+        List<Product> products = new ArrayList<>();
+        try (Connection connection = ConnectionUtil.getConnection()) {
+            PreparedStatement statement = connection.prepareStatement(
+                    "SELECT * FROM products WHERE deleted = 0");
+            ResultSet rs = statement.executeQuery();
+            while (rs.next()) {
+                Product product = getProductFromResultSet(rs);
+                products.add(product);
+            }
+
```
Comment (AnaLanda) [inline src/main/java/internet/shop/dao/jdbc/ProductDaoJdbcImpl.java]: I think you can remove the newline here

**Review by AnaLanda — COMMENTED** [review]
Looks good)

File: src/main/java/internet/shop/dao/jdbc/ProductDaoJdbcImpl.java [hunk 3]
```diff
@@ -0,0 +1,106 @@
...
+            statement.setString(1, object.getName());
+            statement.setDouble(2, object.getPrice());
+            statement.executeUpdate();
+            ResultSet resultSet = statement.getGeneratedKeys();
+            if (resultSet.next()) {
+                object.setId(resultSet.getLong(1));
+            }
+            return object;
+        } catch (SQLException e) {
+            throw new RuntimeException("Can't create product", e);
```
Comment (Rudiachenko) [inline src/main/java/internet/shop/dao/jdbc/ProductDaoJdbcImpl.java]: you should to throw your own exception

File: src/main/java/internet/shop/dao/jdbc/ProductDaoJdbcImpl.java (hunk 1 above)
Comment (jah701) [reply src/main/java/internet/shop/dao/jdbc/ProductDaoJdbcImpl.java]: Agree

Comment (jah701) [reply src/main/java/internet/shop/dao/jdbc/ProductDaoJdbcImpl.java]: Fixed

File: src/main/java/internet/shop/dao/jdbc/ProductDaoJdbcImpl.java (hunk 2 above)
Comment (jah701) [reply src/main/java/internet/shop/dao/jdbc/ProductDaoJdbcImpl.java]: Agree 👍

File: src/main/java/internet/shop/Application.java [hunk 4]
```diff
@@ -1,69 +1,23 @@
...
-        final ShoppingCartService shoppingCartService
-                = (ShoppingCartService) injector.getInstance(ShoppingCartService.class);
-        final ProductService productService
-                = (ProductService) injector.getInstance(ProductService.class);
-        final OrderService orderService = (OrderService) injector.getInstance(OrderService.class);
-        User admin1 = new User("admin1", "admin1", "admin");
-        User admin2 = new User("admin2", "admin2", "admin");
-        User admin3 = new User("admin3", "admin3", "admin");
-        User admin4 = new User("admin4", "admin4", "admin");
+        ProductDaoJdbcImpl productDaoJdbc = new ProductDaoJdbcImpl();
```
Comment (boroda4436) [inline src/main/java/internet/shop/Application.java]:
```suggestion
        ProductDao productDao = new ProductDaoJdbcImpl();
```

File: src/main/java/internet/shop/dao/jdbc/ProductDaoJdbcImpl.java [hunk 5]
```diff
@@ -0,0 +1,106 @@
...
+    @Override
+    public Optional<Product> get(Long id) {
+        Product product = null;
+        try (Connection connection = ConnectionUtil.getConnection()) {
+            PreparedStatement statement = connection.prepareStatement(
+                    "SELECT * FROM products WHERE product_id = ?;");
+            statement.setLong(1, id);
+            ResultSet rs = statement.executeQuery();
+            while (rs.next()) {
+                product = getProductFromResultSet(rs);
```
Comment (boroda4436) [inline src/main/java/internet/shop/dao/jdbc/ProductDaoJdbcImpl.java:47]: return immediately

File: src/main/java/internet/shop/dao/jdbc/ProductDaoJdbcImpl.java [hunk 6]
```diff
@@ -0,0 +1,106 @@
...
+            }
+        } catch (SQLException e) {
+            throw new DataProcessingException("Can't get the product", e);
+        }
+        return Optional.ofNullable(product);
```
Comment (boroda4436) [inline src/main/java/internet/shop/dao/jdbc/ProductDaoJdbcImpl.java]:
```suggestion
        return Optional.ofEmpty();
```

File: src/main/java/internet/shop/dao/jdbc/ProductDaoJdbcImpl.java [hunk 7]
```diff
@@ -0,0 +1,106 @@
...
+            }
+            return object;
+        } catch (SQLException e) {
+            throw new DataProcessingException("Can't create product", e);
+        }
+    }
+
+    @Override
+    public Optional<Product> get(Long id) {
+        Product product = null;
```
Comment (boroda4436) [inline src/main/java/internet/shop/dao/jdbc/ProductDaoJdbcImpl.java]: remember today's lesson

File: src/main/java/internet/shop/dao/jdbc/ProductDaoJdbcImpl.java [hunk 8]
```diff
@@ -0,0 +1,106 @@
...
+        }
+        return Optional.ofNullable(product);
+    }
+
+    @Override
+    public List<Product> getAll() {
+        List<Product> products = new ArrayList<>();
+        try (Connection connection = ConnectionUtil.getConnection()) {
+            PreparedStatement statement = connection.prepareStatement(
+                    "SELECT * FROM products WHERE deleted = 0;");
```
Comment (boroda4436) [inline src/main/java/internet/shop/dao/jdbc/ProductDaoJdbcImpl.java]:
```suggestion
                    "SELECT * FROM products WHERE deleted = false;");
```

File: src/main/java/internet/shop/dao/jdbc/ProductDaoJdbcImpl.java [hunk 9]
```diff
@@ -0,0 +1,106 @@
...
+            throw new DataProcessingException("Can't create product", e);
+        }
+    }
+
+    @Override
+    public Optional<Product> get(Long id) {
+        Product product = null;
+        try (Connection connection = ConnectionUtil.getConnection()) {
+            PreparedStatement statement = connection.prepareStatement(
+                    "SELECT * FROM products WHERE product_id = ?;");
```
Comment (boroda4436) [inline src/main/java/internet/shop/dao/jdbc/ProductDaoJdbcImpl.java]: and deleted = false

File: src/main/java/internet/shop/dao/jdbc/ProductDaoJdbcImpl.java [hunk 10]
```diff
@@ -0,0 +1,106 @@
...
+            throw new DataProcessingException("Can't get all products list", e);
+        }
+        return products;
+    }
+
+    @Override
+    public Product update(Product object) {
+        try (Connection connection = ConnectionUtil.getConnection()) {
+            PreparedStatement statement = connection.prepareStatement(
+                    "UPDATE products SET name = ?, price = ? WHERE product_id = ?;");
```
Comment (boroda4436) [inline src/main/java/internet/shop/dao/jdbc/ProductDaoJdbcImpl.java]: and deleted = false

File: src/main/java/internet/shop/dao/jdbc/ProductDaoJdbcImpl.java [hunk 11]
```diff
@@ -0,0 +1,106 @@
...
+        try (Connection connection = ConnectionUtil.getConnection()) {
+            PreparedStatement statement = connection.prepareStatement(
+                    "SELECT * FROM products WHERE product_id = ?;");
+            statement.setLong(1, id);
+            ResultSet rs = statement.executeQuery();
+            while (rs.next()) {
+                product = getProductFromResultSet(rs);
+            }
+        } catch (SQLException e) {
+            throw new DataProcessingException("Can't get the product", e);
```
Comment (boroda4436) [inline src/main/java/internet/shop/dao/jdbc/ProductDaoJdbcImpl.java]:
```suggestion
            throw new DataProcessingException("Can't get the product by id = " + id, e);
```

File: src/main/java/internet/shop/dao/jdbc/ProductDaoJdbcImpl.java [hunk 12]
```diff
@@ -0,0 +1,106 @@
...
+import java.sql.SQLException;
+import java.sql.Statement;
+import java.util.ArrayList;
+import java.util.List;
+import java.util.Optional;
+
+@Dao
+public class ProductDaoJdbcImpl implements ProductDao {
+    @Override
+    public Product create(Product object) {
```
Comment (boroda4436) [inline src/main/java/internet/shop/dao/jdbc/ProductDaoJdbcImpl.java]:
```suggestion
    public Product create(Product product) {
```

File: src/main/java/internet/shop/dao/jdbc/ProductDaoJdbcImpl.java [hunk 13]
```diff
@@ -0,0 +1,106 @@
...
+        } catch (SQLException e) {
+            throw new DataProcessingException("Can't update product", e);
+        }
+    }
+
+    @Override
+    public boolean delete(Long id) {
+        try (Connection connection = ConnectionUtil.getConnection()) {
+            PreparedStatement statement = connection.prepareStatement(
+                    "UPDATE products SET deleted = 1 WHERE product_id = ?;");
```
Comment (boroda4436) [inline src/main/java/internet/shop/dao/jdbc/ProductDaoJdbcImpl.java]: see Sofia's message in the slack

Comment (boroda4436) [inline src/main/java/internet/shop/dao/jdbc/ProductDaoJdbcImpl.java]: and deleted = false;

File: src/main/java/internet/shop/dao/jdbc/ProductDaoJdbcImpl.java [hunk 14]
```diff
@@ -0,0 +1,106 @@
...
+            statement.setString(1, String.valueOf(id));
+            int i = statement.executeUpdate();
```
Comment (boroda4436) [inline src/main/java/internet/shop/dao/jdbc/ProductDaoJdbcImpl.java]:
```suggestion
            return statement.executeUpdate() == 1;
```

File: src/main/java/internet/shop/util/ConnectionUtil.java [hunk 15]
```diff
@@ -0,0 +1,30 @@
...
+            Class.forName("com.mysql.jdbc.Driver");
+        } catch (ClassNotFoundException e) {
+            throw new RuntimeException("Can't find mysql Driver", e);
+        }
+    }
+
+    public static Connection getConnection() {
+        Properties dbProperties = new Properties();
+        dbProperties.put("user", "root");
+        dbProperties.put("password", "p81gORq76");
```
Comment (boroda4436) [inline src/main/java/internet/shop/util/ConnectionUtil.java:20]: never push your real credentials to github

**Review by boroda4436 — CHANGES_REQUESTED** [review]

File: src/main/java/internet/shop/dao/jdbc/ProductDaoJdbcImpl.java (hunk 7 above)
Comment (jah701) [reply src/main/java/internet/shop/dao/jdbc/ProductDaoJdbcImpl.java]: Тут из за области видимости пришлось так сделать.

File: src/main/java/internet/shop/util/ConnectionUtil.java (hunk 15 above)
Comment (jah701) [reply src/main/java/internet/shop/util/ConnectionUtil.java:20]: yeah ive got it too late :D

File: src/main/java/internet/shop/Application.java (hunk 4 above)
Comment (jah701) [reply src/main/java/internet/shop/Application.java]: fixed

File: src/main/java/internet/shop/dao/jdbc/ProductDaoJdbcImpl.java (hunk 5 above)
Comment (jah701) [reply src/main/java/internet/shop/dao/jdbc/ProductDaoJdbcImpl.java:47]: done

File: src/main/java/internet/shop/dao/jdbc/ProductDaoJdbcImpl.java (hunk 6 above)
Comment (jah701) [reply src/main/java/internet/shop/dao/jdbc/ProductDaoJdbcImpl.java]: done

File: src/main/java/internet/shop/dao/jdbc/ProductDaoJdbcImpl.java (hunk 12 above)
Comment (jah701) [reply src/main/java/internet/shop/dao/jdbc/ProductDaoJdbcImpl.java]: done

**Review by RomanDubovskyi — APPROVED** [review]

**Comment by RomanDubovskyi** [comment]
fix travis first

**Pull request merged**